File: python/day2.py
from utilities import read_lines
from collections import Counter


class Policy:
    def __init__(self, record):
        parts = record.split(':')
        self.setup_policy(parts[0])
        self.password = parts[1]

    def setup_policy(self, policy):
        parts = policy.split(' ')
        self.letter = parts[1]
        range_ = parts[0].split('-')
        self.low_limit = int(range_[0])
        self.high_limit = int(range_[1])


def validator_1(policy):
    counter = Counter(policy.password)
    return policy.low_limit <= counter[policy.letter] <= policy.high_limit


def validator_2(policy):
    return (policy.password[policy.low_limit] == policy.letter or
            policy.password[policy.high_limit] == policy.letter) and (
        policy.password[policy.low_limit] != policy.password[policy.high_limit]
    )


def validate_policy(line, validator):
    return validator(Policy(line))


def part_1(data):
    return sum(map(lambda line: validate_policy(line, validator_1), data))


def part_2(data):
    return sum(map(lambda line: validate_policy(line, validator_2), data))


if __name__ == '__main__':
    print('Day 2')
    data = read_lines('../inputs/day2.txt')
    print('\tPart 1: {}'.format(part_1(data)))
    print('\tPart 2: {}'.format(part_2(data)))

File: rust/src/day2.rs
use crate::utilities::read_lines;

struct Policy {
    letter: char,
    first: usize,
    second: usize,
    password: Vec<char>,
}

impl Policy {
    fn new(s: &String) -> Policy {
        let parts: Vec<&str> = s.split(':').collect();
        let policy: Vec<&str> = parts[0].split_ascii_whitespace().collect();
        let letter = policy[1].chars().next().unwrap();
        let numbers: Vec<usize> = policy[0]
            .split('-')
            .map(|num| num.parse().unwrap())
            .collect();
        Policy {
            letter,
            first: numbers[0],
            second: numbers[1],
            password: parts[1].chars().collect(),
        }
    }

    fn validate_password_1(&self) -> bool {
        let mut counter = 0;
        for &ch in &self.password {
            if ch == self.letter {
                counter += 1;
            }
        }
        counter >= self.first && counter <= self.second
    }

    fn validate_password_2(&self) -> bool {
        self.password.len() > self.second
            && self.password[self.first] != self.password[self.second]
            && (self.password[self.first] == self.letter
                || self.password[self.second] == self.letter)
    }
}

fn part_1(data: &Vec<Policy>) {
    println!(
        "\tPart 1: {}",
        data.iter()
            .filter(|policy| policy.validate_password_1())
            .count()
    );
}

fn part_2(data: &Vec<Policy>) {
    println!(
        "\tPart 2: {}",
        data.iter()
            .filter(|policy| policy.validate_password_2())
            .count()
    );
}

pub fn main() {
    println!("Day 2");
    let data: Vec<Policy> = read_lines("../inputs/day2.txt")
        .iter()
        .map(Policy::new)
        .collect();
    part_1(&data);
    part_2(&data);
}
